Redesign Dashboard Naming Interaction

alexpaxton opened the pull request:
- [x] CHANGELOG.md updated with a link to the PR (not the Issue)
  - [x] Rebased/mergable
  - [x] Tests pass
  - [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

Connect #1991 
Connect #2064 
Connect #2097 

### The Problem
- Running out of space in the Dashboard page header
- Renaming a dashboard does not automatically select text in input (unexpected behavior)

### The Solution
- Remove `Rename` button
- Add dashboard switching dropdown next to Dashboard name (preserve switching functionality)
- Click on dashboard name to rename
- `ESC` cancels rename
- `Enter` or `Blur` saves rename
- Maximum length of dashboard name is 50 characters (based on `1440x900` screen size)
- Dashboard name input automatically selects all text on edit

### Preview
![screen shot 2017-10-10 at 3 19 35 pm](https://user-images.githubusercontent.com/2433762/31413819-0f99c5de-add0-11e7-82c5-dbd791dbe4c7.png)
More space efficient

![dash-switcher](https://user-images.githubusercontent.com/2433762/31413861-35616d26-add0-11e7-8046-e4211589e9e2.gif)
Switching between dashboards (is alphabetically sorted as well)

![dash-rename-4](https://user-images.githubusercontent.com/2433762/31747669-ddf3b73e-b422-11e7-8fe0-d42015674a15.gif)
Click to rename dashboard

### Design Considerations
- This fundamentally changes how users will accomplish 2 tasks, will add a learning curve for existing users
- Users previously clicked on the dashboard name to switch
  - Might accidentally discover how renaming works
  - New means of renaming is in the exact same spot
- Using the `text` cursor on the dashboard name to indicate it can be edited
- Added a `Pencil` icon to the right of the name that highlights on hover, also indicating editability 
- This renaming pattern is inconsistent with other parts of the UI
  - Works well in this instance, going to try it out and if it goes well might re-use elsewhere or update to match other patterns

alexpaxton commented:
Might have a couple very small adjustments left to do

alexpaxton commented:
@121watts made the changes you requested, also undid the pass-menu-items-in-as-children thing cuz it was weird

121watts commented on ui/src/dashboards/components/DashboardHeader.js:
```diff
@@ -5,46 +5,56 @@ import AutoRefreshDropdown from 'shared/components/AutoRefreshDropdown'
+              dashboards={dashboards}
+              currentDashboard={dashboardName}
+              sourceID={sourceID}
+            />
```
👏 😸

121watts commented on ui/src/dashboards/components/DashboardSwitcher.js:
```diff
@@ -0,0 +1,75 @@
+          <span className="icon dash-f" />
+        </button>
+        <ul className="dropdown-menu">
+          {dashboards.map((d, i) =>
```
If you sort these this will resolve issue #1991.

121watts commented on ui/src/dashboards/components/DashboardSwitcher.js:
```diff
@@ -0,0 +1,75 @@
+
+    if (dashboards.length <= 1) {
+      return null
+    }
```
The parent should be in charge of deciding whether or not to render its child.

alexpaxton commented:
@121watts made all the requested changes. I also removed `isRequired` from a few props that are not passed into the `HostPage` dashboards. I also refactored `HostPage` to use the latest patterns and not use children. Not sure if I should be guarding against missing props with these two
